api/internal/localizer/locloader.go

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package localizer
+
+import (
+	"log"
+	"path/filepath"
+
+	"sigs.k8s.io/kustomize/api/ifc"
+	"sigs.k8s.io/kustomize/api/internal/git"
+	"sigs.k8s.io/kustomize/api/loader"
+	"sigs.k8s.io/kustomize/kyaml/errors"
+	"sigs.k8s.io/kustomize/kyaml/filesys"
+)
+
+const dstPrefix = "localized"
+
+// LocLoader is the Loader for kustomize localize. It Loads under localize constraints.
+type LocLoader struct {
+	fSys filesys.FileSystem
+
+	// directory that bounds target's local references, empty string if target is remote
+	scope filesys.ConfirmedDir
+
+	// localize destination
+	newDir filesys.ConfirmedDir
+
+	// loader at LocLoader's current directory
+	ifc.Loader
+
+	// whether LocLoader and all its ancestors are the result of local references
+	local bool
+}
+
+// ValidateLocArgs validates the arguments to kustomize localize.
+// On success, ValidateLocArgs returns a LocLoader at targetArg and the localize destination directory; otherwise, an
+// error.
+func ValidateLocArgs(targetArg string, scopeArg string, newDirArg string, fSys filesys.FileSystem) (*LocLoader, filesys.ConfirmedDir, error) {
+	// check earlier to avoid cleanup
+	spec, err := git.NewRepoSpecFromURL(targetArg)
+	if err == nil && spec.Ref == "" {
+		return nil, "", errors.Errorf("localize remote root '%s' missing ref query string parameter", targetArg)
+	}
+
+	// for security, should enforce load restrictions
+	ldr, err := loader.NewLoader(loader.RestrictionRootOnly, targetArg, fSys)
+	if err != nil {
+		return nil, "", errors.WrapPrefixf(err, "unable to establish localize target '%s'", targetArg)
+	}
+
+	scope, err := validateScope(scopeArg, targetArg, ldr, fSys)
+	if err != nil {
+		_ = ldr.Cleanup()
+		return nil, "", errors.WrapPrefixf(err, "invalid localize scope '%s'", scopeArg)
+	}
+
+	newDir, err := validateNewDir(newDirArg, ldr, spec, fSys)
+	if err != nil {
+		_ = ldr.Cleanup()
+		return nil, "", errors.WrapPrefixf(err, "invalid localize destination '%s'", newDirArg)
+	}
+
+	return &LocLoader{
+		fSys:   fSys,
+		scope:  scope,
+		newDir: newDir,
+		Loader: ldr,
+		local:  scope != "",
+	}, newDir, nil
+}
+
+// Scope returns the localize scope
+func (ll *LocLoader) Scope() string {
+	return ll.scope.String()
+}
+
+// Load returns the contents of path if path is a valid localize file.
+// Otherwise, Load returns an error.
+func (ll *LocLoader) Load(path string) ([]byte, error) {
+	// checks in root, and thus in scope
+	content, err := ll.Loader.Load(path)
+	if err != nil {
+		return nil, errors.WrapPrefixf(err, "invalid file reference")
+	}
+	if filepath.IsAbs(path) {
+		return nil, errors.Errorf("absolute paths not yet supported in alpha: file path '%s' is absolute", path)
+	}
+
+	abs := filepath.Join(ll.Root(), path)
+	dir, f, err := ll.fSys.CleanedAbs(abs)
+	if err != nil {
+		// should never happen
+		log.Fatalf(errors.WrapPrefixf(err, "cannot clean validated file path '%s'", abs).Error())
```

Comment:
As a general rule, we return errors rather than logging them.

Reply:
Understood, but I think this is a special case. When we encounter an error that should never be thrown, we don't know how to proceed. This means there is a bug in the code, in which case we should log it and fail immediately. If we return this error to the upper layer, the upper layer will mistake it for an expected error and incorrectly proceed with other alternatives.

Referenced code snippets like [these](https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/filesys/fsondisk.go#L69)

Reply:
Ah! Wasn't aware that there are other places that we choose to log. This is fine then, thanks!